src/imperative/imperative_utils.h

```diff
@@ -164,7 +164,12 @@ inline void SetShapeType(const Context& ctx,
     NDArrayStorageType storage_type = static_cast<NDArrayStorageType>(out_storage_types[i]);
     if (outputs[i]->is_none()) {
       if (storage_type == kDefaultStorage) {
-        *outputs[i] = NDArray(out_shapes[i], ctx, true, out_types[i]);
+        if (outputs.size() == inputs.size()) {
```

Comment:
???

Reply:
This is for AllReduce operator, the outputs should be in the same device as the corresponding input.

Reply:
This is too hacky

Reply:
I aggree. I am not familiar with very backend code. Could you give suggestions?